regcomp.c: Fix named sequences in (?[...])

khwilliamson opened the pull request:
The regex_sets feature cannot yet handle named sequences possibly
returned by \p{name=...}.  I forgot to check for this possibility which
led to a null pointer dereference.  Also, the called function was
returning success when it should have failed in this circumstance.

This fixes #17732

khwilliamson committed:
regcomp.c: Fix named sequences in (?[...])

The regex_sets feature cannot yet handle named sequences possibly
returned by \p{name=...}.  I forgot to check for this possibility which
led to a null pointer dereference.  Also, the called function was
returning success when it should have failed in this circumstance.

This fixes #17732

hvds commented:
regclass() is now over 2700 lines long, it becomes rather hard to grok the possible paths through it.

The return at line 18591 (the `if (multi_char_matches)` block) does not appear to set ret_invlist.

I'm working to try and analyse the rest of it, it's slow going.

khwilliamson commented:
On 4/27/20 9:16 AM, Hugo van der Sanden wrote:
> regclass() is now over 2700 lines long, it becomes rather hard to grok 
> the possible paths through it.

I followed my predecessors, so that mostly, code that is only used once 
is kept in-line.  But this function is way too unwieldy.  So yes, we 
should start splitting out portions.  The 'if (optimizable) { ... }' 
code is 1000 lines, for example.
> 
> The return at line 18591 (the |if (multi_char_matches)| block) does not 
> appear to set ret_invlist.

I think that is ok as-is for this release.  ret_invlist is used only 
when this is called from (?[...]), and multi-char returns are not 
accepted for that, so this code shouldn't be getting called.  Adding a 
line like

if (allow_multiple_chars && ret_invlist) panic()

should not cause failure.  Going forward, in keeping with the new 
Unicode guidelines for regexes, I will remove these restrictions.

There's three ways multi-chars could get generated:

1) \N{named sequence}.  That is forbidden by this (poorly worded) error 
message
       vFAIL("\\N{} here is restricted to one character");

2) By \p{name=named sequence} (or a wildcard to that effect).  That is 
where I screwed up and forgot to forbid, until the branch you are reviewing

3) By /[\xDF]/i.  That is forbidden by the parameters.  I just tested 
it.  It should give a warning, but doesn't.  Too late for this release, 
and next release should not have this restriction

> 
> I'm working to try and analyse the rest of it, it's slow going.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/Perl/perl5/pull/17751#issuecomment-620050513>, or 
> unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AAA2DH2Z2B5TDRNJOJUFO3TROWOTVANCNFSM4MRHJJTA>.
>

hvds commented:
> Adding a line like `if (allow_multiple_chars && ret_invlist) panic()` should not cause failure.

That seems like a good option; I'd be happy with this given that addition.

khwilliamson committed:
regcomp.c: Add assert

The code's proper functioning currently depends on bracketed character
classes not returning a sequence of characters to the regex sets feature
(?[...])

Add an assert to validate that.

hvds commented:
> > Adding a line like `if (allow_multiple_chars && ret_invlist) panic()` should not cause failure.
> 
> That seems like a good option; I'd be happy with this given that addition.

Ah, the assert commited looks before the fact at allow_multiple_chars where I had expected it to look after the fact at multi_char_matches.

multi_char_matches is set at lines 17620 and 17780 in blocks not obviously protected by allow_multiple_chars.

khwilliamson commented:
On 4/28/20 12:41 PM, Hugo van der Sanden wrote:
>         Adding a line like |if (allow_multiple_chars && ret_invlist)
>         panic()| should not cause failure.
> 
>     That seems like a good option; I'd be happy with this given that
>     addition.
> 
> Ah, the assert commited looks before the fact at allow_multiple_chars 
> where I had expected it to look after the fact at multi_char_matches.
> 
> multi_char_matches is set at lines 17620 and 17780 in blocks not 
> obviously protected by allow_multiple_chars.

I can add something in those places, but it doesn't seem necessary to me.

The else clause containing your first example is executed only if 
allow_multiple chars is set, because the 'if' clause's first line is
      if ( ! allow_mutiple_chars
so that protects it.

Your second example is in a while loop in an else clause, that is 
executed only if the 'if' clause isn't.  And that 'if' tests for 
ret_invlist not being NULL.  So the set is protected by ret_invlist, 
which that assert ties with allow_multiple_chars.

What would make you happy here?

hvds commented:
@khwilliamson let me check again, it's possible I've misread it.

hvds commented:
> The else clause containing your first example is executed only if allow_multiple chars is set, because the 'if' clause's first line is if ( ! allow_mutiple_chars so that protects it.

Yes, I somehow missed that, I'm fine with this one. The misspelling of 'multiple' in `allow_mutiple_chars` might have helped me miss it, not sure.

> Your second example is in a while loop in an else clause, that is executed only if the 'if' clause isn't. And that 'if' tests for ret_invlist not being NULL. So the set is protected by ret_invlist, which that assert ties with allow_multiple_chars.

Ah, that test was added within the branch; for the last check I'd been looking at (roughly) blead, sorry.

So I no longer see any current bug. I look forward to future decomplexification of this function.
